server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java

```diff
@@ -52,101 +48,51 @@ public void preProcess(SearchContext context) {
 
     @Override
     public void execute(SearchContext context) {
-        final ObjectHashSet<Term> termsSet = new ObjectHashSet<>();
         try {
-            context.searcher().createWeight(context.searcher().rewrite(context.query()), ScoreMode.COMPLETE, 1f)
-                .extractTerms(new DelegateSet(termsSet));
-            for (RescoreContext rescoreContext : context.rescore()) {
-                try {
-                    rescoreContext.rescorer().extractTerms(context.searcher(), rescoreContext, new DelegateSet(termsSet));
-                } catch (IOException e) {
-                    throw new IllegalStateException("Failed to extract terms", e);
-                }
-            }
-
-            Term[] terms = termsSet.toArray(Term.class);
-            TermStatistics[] termStatistics = new TermStatistics[terms.length];
-            IndexReaderContext indexReaderContext = context.searcher().getTopReaderContext();
-            for (int i = 0; i < terms.length; i++) {
-                if(context.isCancelled()) {
-                    throw new TaskCancelledException("cancelled");
-                }
-                // LUCENE 4 UPGRADE: cache TermStates?
-                TermStates termContext = TermStates.build(indexReaderContext, terms[i], true);
-                termStatistics[i] = context.searcher().termStatistics(terms[i], termContext);
-            }
-
             ObjectObjectHashMap<String, CollectionStatistics> fieldStatistics = HppcMaps.newNoNullKeysMap();
-            for (Term term : terms) {
-                assert term.field() != null : "field is null";
-                if (fieldStatistics.containsKey(term.field()) == false) {
-                    final CollectionStatistics collectionStatistics = context.searcher().collectionStatistics(term.field());
-                    if (collectionStatistics != null) {
-                        fieldStatistics.put(term.field(), collectionStatistics);
```

Comment:
I wonder if this is really equivalent. Some queries are going to build term statistics even though they don't add terms in `extractTerms` for various reasons (`GlobalOrdinalsQuery` for instance) so we'll end up with more terms than before. However the good side of this change is that we'd extract only the terms that are used for scoring instead of all terms that are present in the query.
The API change for the rescorer is too obfuscated IMO but one thing I fully agree with is that we don't need the special map so we could rely on plain `HashMap` to cleanup the code a bit ?

Reply:
> Some queries are going to build term statistics even though they don't add terms in `extractTerms` for various reasons (`GlobalOrdinalsQuery` for instance)

I think this is incorrect - queries will only build term statistics if they actually need them, whereas `extractTerms` doesn't always track that.  `TermWeight` for example always adds its term, even if it has been created with `ScoreMode.COMPLETE_NO_SCORES`, while `GlobalOrdinalsQuery` creates its child weight with `COMPLETE_NO_SCORES` so no term stats will be pulled here.

I agree about the API change on rescorer though, let me think about a better way to do that.